Couchbase Lite Android 2.0.0-DB018 / delete() and close() fail after a bulk storage with inBatch()

Let me start this issue by indicating that I'm new to both couchbase and couchbase lite. What I would like to do is delete all documents in the database. I'd appreciate some method like database.clear() or database.deleteAllDocuments(). But these methods don't exist.

So as a workaround, I tried to drop the database. I found 2 methods that could do the job close() and delete(), but their respective documentation aren't very clear on their differences. I assumed then that :
- close() should be used for releasing a session on the database
- delete() actually deletes the data by deleting the file on the disk

Hence I tried to call these methods in different ways. Some of them work, but I have a scenario that doesn't work, and I can't get to figure out why.

I do the following : 
- I create a db instance, that a I delete and recreate right away to "clear" it.
- I then add some bulk data using inBatch()  (50000 objects) then attempt to close the db
And that's pretty much it. 
What actually happens at runtime :
- On the first execution, creation, deletion and recreation works but the call to close() throws a CouchbaseLiteException _Failed in closing database transaction not closed_
- On the second run, the call to delete() throws a Failed in CouchbaseLiteException _batch request database busy/locked_

I thought that the second exception on delete() was a consequence of the first exception on close(), but it isn't the case: if I remove the call to close() I still get the same exception on delete(). 
I want to add that I don't understand the message _Failed in closing database transaction not closed_ since it is stated [in the documentation](https://developer.couchbase.com/documentation/mobile/2.0/guides/couchbase-lite/native-api/document/index.html) that transactions don't exist on Couchbase lite 2.0 :

> Couchbase Lite does not offer transactional guarantees, and that the purpose of the method (inBatch(), ndlr) is to optimize batch operations rather than to enable ACID transactions

Can you help ? That is by fixing this bug, or better by providing a deleteAllDocuments() method on the Database class :p

Below is the code and the details of both exceptions. 

**An intent service to bulk store data**
```
package org.nla.testcouch2.couch;

import android.app.IntentService;
import android.content.Intent;
import android.support.annotation.Nullable;
import android.util.Log;

import com.couchbase.lite.CouchbaseLiteException;
import com.couchbase.lite.Database;
import com.couchbase.lite.DatabaseConfiguration;
import com.couchbase.lite.Document;

import java.util.TimerTask;

import org.nla.testcouch2.App;
import org.nla.testcouch2.Constants;
import org.nla.testcouch2.Person;

public class StoreToCouchbaseIntentService extends IntentService {
    public static final String action = "org.nla.testcouch2.couch.StoreToCouchbaseIntentService";
    private Database database = null;

    public StoreToCouchbaseIntentService() {
        super("StoreToCouchbaseIntentService");
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        try {
            database = new Database("application", new DatabaseConfiguration(getApplicationContext()));
            database.delete();
            database = new Database("application", new DatabaseConfiguration(getApplicationContext()));

            long tic = System.currentTimeMillis();
            Log.d("mobiledatabaseperftests", "Starting storage to couch");

            database.inBatch(new TimerTask() {
                @Override
                public void run() {
                    for (Person person : App.app().getPersons()) {
                        Document newPerson = new Document();
                        newPerson.setString(Constants.TYPE, Constants.TYPE_PERSON);
                        newPerson.setInt(Constants.AGE, person.getAge());
                        newPerson.setString(Constants.UUID, person.getGuid());
                        newPerson.setString(Constants.COMPANY, person.getCompany());
                        newPerson.setString(Constants.EMAIL, person.getEmail());
                        newPerson.setString(Constants.EYE_COLOR, person.getEyeColor());
                        newPerson.setString(Constants.PHONE, person.getPhone());
                        newPerson.setString(Constants.PICTURE, person.getPicture());
                        newPerson.setString(Constants.NAME, person.getName());
                        try {
                            database.save(newPerson);
                        } catch (CouchbaseLiteException e) {
                            Log.e("mobiledatabaseperftests", "Failed in storing document " + e.getMessage(), e);
                        }
                    }
                    Log.d("mobiledatabaseperftests", "Batch storage over");
                    try {
                        database.close();
                    } catch (CouchbaseLiteException e) {
                        Log.e("mobiledatabaseperftests", "Failed in closing database " + e.getMessage(), e);
                    }
                    long duration = System.currentTimeMillis() - tic;
                    onBatchTerminated(duration);
                }
            });
        } catch (CouchbaseLiteException e) {
            Log.e("mobiledatabaseperftests", "Failed in  batch request " + e.getMessage(), e);
        }
    }

    private void onBatchTerminated(long duration) {
        Intent broadcastOfflineIntent = new Intent();
        broadcastOfflineIntent.setAction(action);
        broadcastOfflineIntent.addCategory(Intent.CATEGORY_DEFAULT);
        broadcastOfflineIntent.putExtra(Constants.keySuccess, true);
        broadcastOfflineIntent.putExtra(Constants.keyDuration, duration);
        Log.d("mobiledatabaseperftests", "Stored to couchbase offline: " + duration);
        sendBroadcast(broadcastOfflineIntent);
    }
}
```

**Failed in  batch request database busy/locked**
```
Failed in  batch request database busy/locked
CouchbaseLiteException{domain=1, code=24, msg=database busy/locked}
   at com.couchbase.lite.internal.bridge.LiteCoreBridge.convertException(LiteCoreBridge.java:22)
   at com.couchbase.lite.Database.deleteC4DB(Database.java:894)
   at com.couchbase.lite.Database.delete(Database.java:442)
   at org.nla.testcouch2.couch.StoreToCouchbaseIntentService.onHandleIntent(StoreToCouchbaseIntentService.java:33)
   at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:68)
   at android.os.Handler.dispatchMessage(Handler.java:105)
   at android.os.Looper.loop(Looper.java:164)
   at android.os.HandlerThread.run(HandlerThread.java:65)
Caused by: LiteCoreException{domain=1, code=24, msg=database busy/locked}
   at com.couchbase.litecore.C4Database.delete(Native Method)
   at com.couchbase.litecore.C4Database.delete(C4Database.java:56)
   at com.couchbase.lite.Database.deleteC4DB(Database.java:892)
   at com.couchbase.lite.Database.delete(Database.java:442) 
   at org.nla.testcouch2.couch.StoreToCouchbaseIntentService.onHandleIntent(StoreToCouchbaseIntentService.java:33) 
   at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:68) 
   at android.os.Handler.dispatchMessage(Handler.java:105) 
   at android.os.Looper.loop(Looper.java:164) 
   at android.os.HandlerThread.run(HandlerThread.java:65)
```

**Failed in closing database transaction not closed**
```
Failed in closing database transaction not closed
CouchbaseLiteException{domain=1, code=26, msg=transaction not closed}
   at com.couchbase.lite.internal.bridge.LiteCoreBridge.convertException(LiteCoreBridge.java:22)
   at com.couchbase.lite.Database.closeC4DB(Database.java:886)
   at com.couchbase.lite.Database.close(Database.java:424)
   at org.nla.testcouch2.couch.StoreToCouchbaseIntentService$1.run(StoreToCouchbaseIntentService.java:61)
   at com.couchbase.lite.Database.inBatch(Database.java:306)
   at org.nla.testcouch2.couch.StoreToCouchbaseIntentService.onHandleIntent(StoreToCouchbaseIntentService.java:39)
   at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:68)
   at android.os.Handler.dispatchMessage(Handler.java:105)
   at android.os.Looper.loop(Looper.java:164)
   at android.os.HandlerThread.run(HandlerThread.java:65)
Caused by: LiteCoreException{domain=1, code=26, msg=transaction not closed}
   at com.couchbase.litecore.C4Database.close(Native Method)
   at com.couchbase.litecore.C4Database.close(C4Database.java:52)
   at com.couchbase.lite.Database.closeC4DB(Database.java:884)
   at com.couchbase.lite.Database.close(Database.java:424) 
   at org.nla.testcouch2.couch.StoreToCouchbaseIntentService$1.run(StoreToCouchbaseIntentService.java:61) 
   at com.couchbase.lite.Database.inBatch(Database.java:306) 
   at org.nla.testcouch2.couch.StoreToCouchbaseIntentService.onHandleIntent(StoreToCouchbaseIntentService.java:39) 
   at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:68) 
   at android.os.Handler.dispatchMessage(Handler.java:105) 
   at android.os.Looper.loop(Looper.java:164) 
   at android.os.HandlerThread.run(HandlerThread.java:65) 
```

--- a/android/CouchbaseLite/src/androidTest/java/com/couchbase/lite/DatabaseTest.java
+++ b/android/CouchbaseLite/src/androidTest/java/com/couchbase/lite/DatabaseTest.java
@@ -1414,4 +1414,46 @@ public class DatabaseTest extends BaseTest {
         db.deleteIndex("index3");
 
     }
+
+    // https://github.com/couchbase/couchbase-lite-android/issues/1416
+    @Test
+    public void testDeleteAndOpenDB() throws CouchbaseLiteException {
+        DatabaseConfiguration config = new DatabaseConfiguration(this.context);
+        config.setDirectory(dir);
+
+        // open "application" database
+        final Database database1 = new Database("application", config);
+
+        // delete "application" database
+        database1.delete();
+
+        // open "application" database again
+        final Database database2 = new Database("application", config);
+
+        // inserting documents
+        database2.inBatch(new Runnable() {
+            @Override
+            public void run() {
+                // just create 100 documents
+                for (int i = 0; i < 1000; i++) {
+                    Document doc = new Document();
+
+                    // each doc has 10 items
+                    doc.setInt("index", i);
+                    for (int j = 0; j < 9; j++)
+                        doc.setInt("item_" + j, j);
+
+                    try {
+                        database2.save(doc);
+                    } catch (CouchbaseLiteException e) {
+                        Log.e(TAG, "Failed in storing document " + e.getMessage(), e);
+                        fail();
+                    }
+                }
+            }
+        });
+
+        // close db again
+        database2.close();
+    }
 }

